tensorflow/compiler/xla/service/gpu/fusion_bitcast_lift.cc

```diff
@@ -0,0 +1,205 @@
+	  } else if (i->opcode() == HloOpcode::kParameter &&
+		     absl::c_all_of(i->users(), [](HloInstruction* u) {
+                         return u->opcode() == HloOpcode::kBitcast;
+                       })) {
+	    // Replace the parameter inside the fusion.
+	    Shape new_shape = i->users()[0]->shape();
+	    int64 parameter_number = i->parameter_number();
+	    string name = i->name();
+	    auto n = HloInstruction::CreateParameter(parameter_number,
+						     new_shape, name);
+	    HloInstruction* new_parameter =
+	      i->parent()->ReplaceParameter(parameter_number,
+					    std::move(n));
+	    // Remove the old inner bitcast.
+	    auto old_users = new_parameter->users();
+	    for (HloInstruction* param_user : old_users) {
+	      DCHECK(param_user->opcode() == HloOpcode::kBitcast)
+		<< "Expected a bitcast";
+	      param_user->parent()->ReplaceInstructionWithDifferentShape(
+									 param_user, new_parameter);
+	    }
+	    // Replace the corresponding fusion operands with a new bitcast.
+	    HloInstruction* old_outer_parameter =
+	      cloned_fusion->mutable_operand(parameter_number);
+	    HloInstruction* new_op =
+	      old_outer_parameter->parent()->AddInstruction(
+                    HloInstruction::CreateBitcast(new_shape,
+						  old_outer_parameter));
+	    cloned_fusion->ReplaceOperandWithDifferentShape(
+                parameter_number, new_op);
+	    clone_changed = true;
+	    changed = true;
+	  } else if (i->opcode() == HloOpcode::kBroadcast) {
+	    // For now, do nothing. Later we can merge the broadcast
+	    // and the bitcast, but this doesn't bring benefit in my
+	    // current case.
+	    stack.push_back(i->mutable_operand(0));
+	  } else if (i->users().size() > 0 &&
+		     absl::c_all_of(i->users(), [](HloInstruction* u) {
+                         return u->opcode() == HloOpcode::kBitcast;
+                       })) {
+	    // All users are bitcast, so lift the bitcast.
+	    Shape new_shape = i->users()[0]->shape();
+	    std::vector<HloInstruction*> new_operands;
+	    for (HloInstruction* opnd : i->operands()) {
+	      Shape dtyped_new_shape = ShapeUtil::ChangeElementType(
+                  new_shape, opnd->shape().element_type());
+	      HloInstruction* new_opnd = opnd->parent()->AddInstruction(
+                  HloInstruction::CreateBitcast(dtyped_new_shape, opnd));
+	      new_operands.push_back(new_opnd);
+	      // Handle the operand right before the inserted bitcast now.
+	      if (std::find(stack.begin(), stack.end(), opnd) ==
+		  stack.end()) {
+		stack.push_back(opnd);
+	      }
+	    }
+	    Shape dtyped_new_shape = ShapeUtil::ChangeElementType(
+                new_shape, i->shape().element_type());
+	    HloInstruction* cloned_i = i->parent()->AddInstruction(
+                i->CloneWithNewOperands(dtyped_new_shape, new_operands));
+	    // Replace the old bitcasts with the new instruction to
+	    // remove it.
+	    for (HloInstruction* user: i->users()) {
```

Comment:
This seems to be the source of the issue that ASAN complains about. ASAN stops complaining if you copy i->users() to a temporary vector and iterate over that. I assume this is because if you replace user with cloned_i, i->users() would be modified as well which causes an issue when iterating over it.